packages/ddp-client/client/queueStubsHelpers.js

```diff
@@ -104,6 +104,15 @@ export const loadAsyncStubHelpers = () => {
               finished = true;
               reject(err);
             });
+          applyAsyncPromise.serverPromise
+              .then((result) => {
+                finished = true;
+                resolve(result);
+              })
+              .catch((err) => {
+                finished = true;
+                reject(err);
+              });
```

Comment:
This only needs to run after the stub finishes (or immediately if the method does not have a stub). Running it after receiving the server result is too late.

Reply:
Trying to understand better this part of the code and your input is determinant. Thank you.

I was able to run the tests from `ddp-client` package. And I am finally done with a fix that not only ensures the current tests and expectations are still passing, but also the use case that is missing also works as it does in 2.x.

So, as far as I understand, the `applyAsyncPromise` is a promise that is like a combination of stub and server promises. So  in this part of the code, we listen and resolve or reject according to the success or error of the promise combination, any rejection will cause the client operation to not be executed, like the error on this regression `"Method insertAsync not found on the server [404]"`.

But when we have an operation that uses stub (`returnStubValue`) and server (`returnServerResultPromise`), we still want the client operation to run ok, even if the server fails for any reason. So in this scenario, `applyAsyncPromise.stubPromise` is the only promise I care to return and to continue sucessfully with the client collection operation. In any other case, `applyAsyncPromise` combination promise is behaving as expected as I check on the tests.

Still trying to know each part of this code having a better mental picture of this implementation, so any review, correction or feedback to understand further is appracieted.

Reply:
With this change I break other tests like this [one](https://github.com/meteor/meteor/blob/release-3.0/packages/mongo/mongo_livedata_tests.js#L227-L269) that actually expects to handle properly any collection operation server error.

So, in Meteor 2.x, the data is created locally when running collection operations that fail on the server due to "not found" method. It feels like wrong behavior, or more difficult to mimic in the current 3.x code without fibers.

In Meteor 3.x, we can't immediately create the local data within a remote collection context knowing that we need to wait for a server promise to response first.

Or there is something I am missing.

This issue was originally reported [here](https://meteor-community.slack.com/archives/CN350MY1G/p1704695310758029). I wonder if we can think in another way to approach this. Like change the pattern on using client collections for tests in a way that these are created as non-remote collections with `connection: null` on being instantiated on-fly and then they will effectively behave as sync collections.